veps/sig-compute/migration-prio-queue.md

```diff
@@ -0,0 +1,121 @@
+`VirtualMachineInstanceMigration` (VMIM) API in KubeVirt. It enables
+controllers to assign priorities to migrations, ensuring system-initiated
+migrations (e.g., node drains, upgrades) take precedence over user-initiated
+ones (e.g., hot plug operations). Running migrations will receive a reserved
+high priority when re-enqueued, and admission webhooks will limit user-set
+priorities to safeguard system operations.
+
+## Motivation
+
+The current migration controller, uses a priority queue where running
+migrations have a priority of 0, and pending migrations re-enqueued due to
+capacity constraints are set to \-100. This ensures running migrations are
+processed first but does not differentiate between system-critical and
+user-initiated pending migrations. As a result, user-initiated migrations
+(e.g., multiple hot plug operations) can overwhelm the system, delaying
+critical operations like node drains or KubeVirt upgrades. This proposal
+introduces a flexible priority system to ensure system operations are
+prioritized while maintaining the precedence of running migrations.
+
+## Goals
+
+- Add a `priority` field to `VirtualMachineInstanceMigrationSpec`.
+- Update the migration controller to order the queue by `priority`.
+- Assign a high, reserved priority (e.g., 1000) to running migrations.
+- Enable controllers to set priorities based on migration type.
+- Restrict user-set priorities via admission webhooks.
+- Ensure backward compatibility.
+
+## Non Goals
+
+- Change how migrations are triggered or processed.
+- Introduce new migration types.
+- Make migrations preemptible.
+
+## User Stories
+
+- As a cluster administrator, I want KubeVirt system-critical operations to proceed without delays from user actions.
+- As a VM owner, I want my hot plug operations processed efficiently, respecting system priorities.
+
+## Repos
+
+* [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+## Design
+
+A `priority` field will be added to the VMIM API, allowing various controllers to
+set priorities for migrations. The migration controller will use this field
+to order migrations in the priority queue, ensuring system-initiated migrations
+are processed before user-initiated ones. Running migrations will be get a reserved
+high priority (e.g., 1000).
+Controllers will set priorities based on the migration type, and admission webhooks will ensure users cannot
+set high priorities, protecting system operations.
+
+### Priority Order
+
+The priority order for migrations is as follows, with higher numerical values indicating higher priority:
+
+- Running Migrations: 1000 (Protected)
+- System-Critical Migrations: 100 - These include urgent tasks like node drains, evacuations, and KubeVirt upgrades. They’re critical for system health and need immediate attention.
+- User-Triggered Operations: 50 -  These are user-initiated changes, like hot plug operations.  They’re time-sensitive but should not affect system stability.
```

Comment:
WDYT about using priority ranges instead of single values? e.g.:
* system-critical: 100-150
* user-triggered: 50-99

This would give us the flexibility to refine priorities later. IOW I guess not all "system-critical" migrations are equal, e.g. a node-level drain caused by resource outage is probably more important than an API initiated drain. Similarly, we could prioritize different user-triggered operations within their own range (hotplugs are less/more important than X Y Z).

Reply:
I agree. And I thought that I already reworked that, but probably didn't submit :/